### ISSUE TITLE
FSDP - Allow to reshard frozen parameters in the backward pass, for parameter-efficient trainings of much larger models

### ISSUE THREAD

**Issue by jingqiao** [pytorch/pytorch#95805]
### Context 
We are working on parameter-efficient finetuning of large language models where 99%+ or 99.9%+ parameters are frozen. That is, we freeze all the Transformer parameters, while only tuning a small % of extra parameters (e.g., in LoRA by using Huggingface PEFT). We further turn on activation checkpointing and cpu offloading because the model is large. 

### Issue
FSDP works in the forward pass (i.e., flat CUDA memory usage after the forward pass of each Transformer block), but fails in the backward pass because of continuous increase of CUDA memory after the backward pass of each block. 

### Reason
After diving deep, we find it is because `_post_backward_hook` is not registered for FlatParameter with `requires_grad=False` in `_register_post_backward_hooks`. Related code [torch/distributed/fsdp/_runtime_utils.py](https://github.com/pytorch/pytorch/blob/5d29b68bbccdbdf222e855aec02ddb64f56ca490/torch/distributed/fsdp/_runtime_utils.py#L1233) 
```
    for handle in handles:
        flat_param = handle.flat_param
        already_registered = hasattr(flat_param, "_post_backward_hook_state")
        if already_registered or not flat_param.requires_grad:
            continue
        ...
        hook_handle = acc_grad.register_hook(
            functools.partial(_post_backward_hook, state, handle)
        )
```
As a result, the FlatParameter of each frozen Transformer block (`requires_grad=False`) is first unsharded before its backward calculation (in the `_pre_backward_hook`), but never got reshareded afterwards (i.e., NO `_post_backward_hook` is triggered). So the CUDA memory continuously increases after each block's backward calculation, which is different from the flat CUDA memory usage in the forward pass of each block (unshard, then reshard)

### Workaround
As a workaround, we set requires_grad=True for all Transformer blocks, but don't include them in the optimizer. This leads to unnecessary CUDA memory for grad, but not for optimizer states. This way, we observe transformer FlatParameter is reshared in the backward pass, and we can increase the model size to a certain extent. 

### Feature Request
We's like to request a feature that allows to reshard frozen parameters (i.e., FlatParameter with `requires_grad=False`) in the backward pass. This is to avoid introducing unnecessary grad for frozen parameters, and make it feasible to finetune even larger models. 

### Why is this important?
This feature is crucial for parameter-efficient finetuning of big models. The model size is significantly limited without being able to reshard the frozen parameters in the backward pass. This feature will be valuable to many researchers to explore research and applications of LLM. 

### A side question
Even in the above workaround, we still observe some continuous CUDA memory increase after the backward pass of each Transformer block. It is hard to debug the backward pass, e.g., to set breakpoints, except for printing logs. Can anyone share some suggestion for efficient debugging of the backward pass? Not much info after searching the Web.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu

**Comment by awgu** [comment]
Thanks for the detailed issue! Let me discuss with the other FSDP developers as well.

The forefront question in my mind is how we can tell when it is safe (with respect to timing) to reshard the frozen parameters. Normally, the post-backward hook running signifies that gradient computation involving the parameters has completed, so it is safe to reshard.

I do not have suggestion for debugging backward pass either :/

**Comment by jingqiao** [comment]
@awgu Thanks for the prompt response. 

I am thinking about two situations:
- All the parameters of a FSDP module are frozen (including the parameters of all its submodules, which might be also wrapped by FSDP -- let's refer them as FSDP submodule below). 
- All the parameters of a FSDP module (except for some of its FSDP submodules) are frozen. 

In the first situation, it might be more straightforward and safe to reshard. In the second situation, is it also safe to reshard the FSDP module's own frozen parameters in its post-backward hook (which is called after all its FSDP submodules have finished calculation including gradients)?

One exception could be that a FSDP module with frozen parameters is called twice (for whatever reason) in the backward pass of the whole model. But in this case, the frozen parameters will be unsharded in the pre-backward hook when the FSDP module is called again; so still safe, right?

I have probably missed some edge cases, but will be glad to learn and get your help on this.

**Comment by awgu** [comment]
Apologies for the delay.

> Even in the above workaround, we still observe some continuous CUDA memory increase after the backward pass of each Transformer block.

I was wondering, could the increase just be from the gradients computed for the "frozen" `FlatParameter`s that you forced to use `requires_grad=True`?

> In the first situation, it might be more straightforward and safe to reshard. In the second situation, is it also safe to reshard the FSDP module's own frozen parameters in its post-backward hook (which is called after all its FSDP submodules have finished calculation including gradients)?

I may be misunderstanding your idea, so please feel free to correct me if I am wrong. My main concern is that there _cannot_ be a post-backward hook for a `FlatParameter` that does not require gradient. Normally, the post-backward hook is registered on the `FlatParameter'`s `AccumulateGrad` function object that runs when its gradient is finalized. However, if the `FlatParameter` does not require gradient, I do not think that `AccumulateGrad` object exists. Registering the hook on the `AccumulateGrad` object gives us the correct timing to reshard.

Off the top of my mind, there may be an "advanced" technique that might address this. However, I am not sure when we will have the bandwidth to try this out.

Here is what I am thinking:
- https://github.com/pytorch/pytorch/pull/86260 added something called a multi-grad hook that takes in a list of tensors and runs the hook only when all tensors in the list have their gradients computed.
- For an FSDP module with a `FlatParameter` with `requires_grad=False`, we register a multi-grad hook on the _input tensors_ of that FSDP module's forward that require gradient. This hook only performs resharding of the `FlatParameter` without any gradient reduction (i.e. reduce-scatter).
- If I am understanding correctly, we can assume that the `FlatParameter`'s original parameters are only needed to compute the gradient with respect to the input tensors. In that case, the `FlatParameter` should only be resharded after it is no longer needed.
- We need the multi-grad hook because there may be multiple input tensors that require gradient, and we do not know which will be the last to compute its gradient.

In short, FSDP normally uses the point when the `FlatParameter` gradient is finalized to reshard it, but since that does not exist for `requires_grad=False`, we can try to use the point when the input tensors' gradients are finalized.

**Comment by awgu** [comment]
@jingqiao Do you have any recommendation for a workload that I can test on (including any instructions to get it set up)?

I want to test out a PR to see if it helps.

**Comment by awgu** [comment]
Marking this as closed due to https://github.com/pytorch/pytorch/pull/101982
